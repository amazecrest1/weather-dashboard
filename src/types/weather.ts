// Weather API response types
export interface DailyWeatherData {
  temperature_2m_max: number[];
  temperature_2m_min: number[];
  temperature_2m_mean: number[];
  precipitation_sum: number[];
  wind_speed_10m_max: number[];
  time: string[];
}

export interface HourlyWeatherData {
  temperature_2m: number[];
  relative_humidity_2m: number[];
  apparent_temperature: number[];
  precipitation: number[];
  pressure_msl: number[];
  wind_speed_10m: number[];
  time: string[];
}

export interface WeatherApiResponse {
  daily?: DailyWeatherData;
  hourly?: HourlyWeatherData;
  daily_units?: Record<string, string>;
  hourly_units?: Record<string, string>;
}

// City type
export interface City {
  id: string;
  name: string;
  country: string;
  state?: string;
  latitude: number;
  longitude: number;
  isCustom?: boolean;
}

// Chart data types
export interface ChartDataPoint {
  date: string;
  value: number;
  label?: string;
}

export interface TemperatureChartData {
  date: string;
  max: number;
  min: number;
  mean: number;
}

export interface MultiParameterChartData {
  time: string;
  [key: string]: number | string;
}

// Filter types
export interface DateRange {
  start: string;
  end: string;
}

export interface WeatherFilters {
  city: City;
  dateRange: DateRange;
  selectedParameters?: string[];
}

// Parameter configuration
export interface ParameterConfig {
  key: string;
  label: string;
  unit: string;
  color: string;
  yAxisId?: string;
}

// API request parameters
export interface WeatherApiParams {
  latitude: number;
  longitude: number;
  start_date: string;
  end_date: string;
  daily?: string;
  hourly?: string;
  timezone?: string;
} 