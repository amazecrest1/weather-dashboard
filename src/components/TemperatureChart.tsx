import React from 'react';
import { LineChart, Line, XAxis, YAxis, Tooltip, Legend, ResponsiveContainer } from 'recharts';
import { TemperatureChartData } from '../types/weather';

interface TemperatureChartProps {
  data: TemperatureChartData[];
  className?: string;
  onClick?: () => void;
}

const TemperatureChart: React.FC<TemperatureChartProps> = ({ data, className = '', onClick }) => {
  return (
    <div className={className}>
      <div className="flex items-center mb-3 sm:mb-4">
        <div className="flex items-center">
          {/* Thermometer icon */}
          <svg className="w-4 h-4 sm:w-5 sm:h-5 text-gray-600 dark:text-gray-400 mr-2" fill="currentColor" viewBox="0 0 24 24">
            <path d="M12 2C10.34 2 9 3.34 9 5v6.5c-1.21.91-2 2.37-2 3.5a5 5 0 0 0 10 0c0-1.13-.79-2.59-2-3.5V5c0-1.66-1.34-3-3-3zm-1 3h2v2h-2V5zm0 3h2v2h-2V8zm0 3h2v2h-2v-2z"/>
          </svg>
          <h3 className="text-base sm:text-lg font-medium text-gray-900 dark:text-orange-200 bg-gradient-to-r from-orange-500 to-red-500 bg-clip-text text-transparent">
            Temperature
          </h3>
        </div>
      </div>
      <div style={{ width: '100%', height: '250px', minHeight: '200px' }}>
        <ResponsiveContainer>
          <LineChart data={data} margin={{ top: 10, right: 20, left: 20, bottom: 20 }}>
            <XAxis 
              dataKey="date" 
              tick={{ fontSize: 10 }}
              axisLine={false}
              tickLine={false}
              tickFormatter={(value) => {
                const date = new Date(value);
                return date.toLocaleDateString('en-US', { 
                  month: 'short', 
                  day: 'numeric' 
                });
              }}
            />
            <YAxis 
              tick={{ fontSize: 10 }}
              axisLine={false}
              tickLine={false}
              label={{ 
                value: 'Temperature (°C)', 
                angle: -90, 
                position: 'insideLeft',
                style: { textAnchor: 'middle', fontSize: 10 },
                offset: -5
              }}
            />
            <Tooltip 
              formatter={(value: number, name: string) => [
                `${value.toFixed(1)}°C`, 
                name.charAt(0).toUpperCase() + name.slice(1)
              ]}
              labelFormatter={(label) => {
                const date = new Date(label);
                return date.toLocaleDateString('en-US', { 
                  weekday: 'long',
                  year: 'numeric',
                  month: 'long',
                  day: 'numeric'
                });
              }}
              contentStyle={{
                backgroundColor: 'var(--tooltip-bg, rgba(255, 255, 255, 0.95))',
                border: 'var(--tooltip-border, 1px solid #e5e7eb)',
                borderRadius: '6px',
                boxShadow: '0 10px 15px -3px rgba(0, 0, 0, 0.1)',
                color: 'var(--tooltip-text, #374151)'
              }}
              cursor={false}
              isAnimationActive={false}
              animationDuration={0}
            />
            <Legend />
            <Line 
              type="monotone" 
              dataKey="max" 
              stroke="#fbbf24"
              strokeWidth={2.5}
              name="Maximum"
              dot={false}
              activeDot={{ r: 4, fill: '#fbbf24' }}
            />
            <Line 
              type="monotone" 
              dataKey="mean" 
              stroke="#fcd34d"
              strokeWidth={2.5}
              name="Average"
              dot={false}
              activeDot={{ r: 4, fill: '#fcd34d' }}
            />
            <Line 
              type="monotone" 
              dataKey="min" 
              stroke="#f59e0b"
              strokeWidth={2.5}
              name="Minimum"
              dot={false}
              activeDot={{ r: 4, fill: '#f59e0b' }}
            />
          </LineChart>
        </ResponsiveContainer>
      </div>
    </div>
  );
};

export default TemperatureChart; 