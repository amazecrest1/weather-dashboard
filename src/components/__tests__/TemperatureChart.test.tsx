import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import TemperatureChart from '../TemperatureChart';
import { TemperatureChartData } from '../../types/weather';

jest.mock('recharts', () => ({
  LineChart: ({ children }: { children: React.ReactNode }) => <div data-testid="line-chart">{children}</div>,
  Line: ({ name }: { name: string }) => <div data-testid={`line-${name}`}>{name}</div>,
  XAxis: () => <div data-testid="x-axis">XAxis</div>,
  YAxis: () => <div data-testid="y-axis">YAxis</div>,
  Tooltip: () => <div data-testid="tooltip">Tooltip</div>,
  Legend: () => <div data-testid="legend">Legend</div>,
  ResponsiveContainer: ({ children }: { children: React.ReactNode }) => (
    <div data-testid="responsive-container">{children}</div>
  ),
}));

describe('TemperatureChart', () => {
  const mockData: TemperatureChartData[] = [
    { date: 'Jan 1', max: 10, min: 2, mean: 6 },
    { date: 'Jan 2', max: 12, min: 4, mean: 8 },
    { date: 'Jan 3', max: 8, min: 1, mean: 4 },
  ];

  it('renders chart title correctly', () => {
    render(<TemperatureChart data={mockData} />);
    expect(screen.getByText('Temperature')).toBeInTheDocument();
  });

  it('renders chart components', () => {
    render(<TemperatureChart data={mockData} />);
    expect(screen.getByTestId('responsive-container')).toBeInTheDocument();
    expect(screen.getByTestId('line-chart')).toBeInTheDocument();
    expect(screen.getByTestId('x-axis')).toBeInTheDocument();
    expect(screen.getByTestId('y-axis')).toBeInTheDocument();
    expect(screen.getByTestId('tooltip')).toBeInTheDocument();
    expect(screen.getByTestId('legend')).toBeInTheDocument();
  });

  it('renders temperature lines', () => {
    render(<TemperatureChart data={mockData} />);
    expect(screen.getByTestId('line-Maximum')).toBeInTheDocument();
    expect(screen.getByTestId('line-Minimum')).toBeInTheDocument();
    expect(screen.getByTestId('line-Average')).toBeInTheDocument();
  });

  it('applies custom className', () => {
    render(<TemperatureChart data={mockData} className="custom-class" />);
    const wrapper = document.querySelector('.custom-class');
    expect(wrapper).toBeInTheDocument();
    expect(wrapper).toHaveClass('custom-class');
  });

  it('renders with empty data', () => {
    render(<TemperatureChart data={[]} />);
    expect(screen.getByText('Temperature')).toBeInTheDocument();
    expect(screen.getByTestId('line-chart')).toBeInTheDocument();
  });

  it('matches snapshot with data', () => {
    const { container } = render(<TemperatureChart data={mockData} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot with empty data', () => {
    const { container } = render(<TemperatureChart data={[]} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot with custom className', () => {
    const { container } = render(<TemperatureChart data={mockData} className="custom-class" />);
    expect(container.firstChild).toMatchSnapshot();
  });
}); 