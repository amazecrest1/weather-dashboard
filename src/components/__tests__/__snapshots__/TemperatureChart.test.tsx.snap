// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TemperatureChart matches snapshot with custom className 1`] = `
<div
  class="custom-class"
>
  <div
    class="flex items-center mb-3 sm:mb-4"
  >
    <div
      class="flex items-center"
    >
      <svg
        class="w-4 h-4 sm:w-5 sm:h-5 text-gray-600 dark:text-gray-400 mr-2"
        fill="currentColor"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 2C10.34 2 9 3.34 9 5v6.5c-1.21.91-2 2.37-2 3.5a5 5 0 0 0 10 0c0-1.13-.79-2.59-2-3.5V5c0-1.66-1.34-3-3-3zm-1 3h2v2h-2V5zm0 3h2v2h-2V8zm0 3h2v2h-2v-2z"
        />
      </svg>
      <h3
        class="text-base sm:text-lg font-medium text-gray-900 dark:text-orange-200 bg-gradient-to-r from-orange-500 to-red-500 bg-clip-text text-transparent"
      >
        Temperature
      </h3>
    </div>
  </div>
  <div
    style="width: 100%; height: 250px; min-height: 200px;"
  >
    <div
      data-testid="responsive-container"
    >
      <div
        data-testid="line-chart"
      >
        <div
          data-testid="x-axis"
        >
          XAxis
        </div>
        <div
          data-testid="y-axis"
        >
          YAxis
        </div>
        <div
          data-testid="tooltip"
        >
          Tooltip
        </div>
        <div
          data-testid="legend"
        >
          Legend
        </div>
        <div
          data-testid="line-Maximum"
        >
          Maximum
        </div>
        <div
          data-testid="line-Average"
        >
          Average
        </div>
        <div
          data-testid="line-Minimum"
        >
          Minimum
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TemperatureChart matches snapshot with data 1`] = `
<div
  class=""
>
  <div
    class="flex items-center mb-3 sm:mb-4"
  >
    <div
      class="flex items-center"
    >
      <svg
        class="w-4 h-4 sm:w-5 sm:h-5 text-gray-600 dark:text-gray-400 mr-2"
        fill="currentColor"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 2C10.34 2 9 3.34 9 5v6.5c-1.21.91-2 2.37-2 3.5a5 5 0 0 0 10 0c0-1.13-.79-2.59-2-3.5V5c0-1.66-1.34-3-3-3zm-1 3h2v2h-2V5zm0 3h2v2h-2V8zm0 3h2v2h-2v-2z"
        />
      </svg>
      <h3
        class="text-base sm:text-lg font-medium text-gray-900 dark:text-orange-200 bg-gradient-to-r from-orange-500 to-red-500 bg-clip-text text-transparent"
      >
        Temperature
      </h3>
    </div>
  </div>
  <div
    style="width: 100%; height: 250px; min-height: 200px;"
  >
    <div
      data-testid="responsive-container"
    >
      <div
        data-testid="line-chart"
      >
        <div
          data-testid="x-axis"
        >
          XAxis
        </div>
        <div
          data-testid="y-axis"
        >
          YAxis
        </div>
        <div
          data-testid="tooltip"
        >
          Tooltip
        </div>
        <div
          data-testid="legend"
        >
          Legend
        </div>
        <div
          data-testid="line-Maximum"
        >
          Maximum
        </div>
        <div
          data-testid="line-Average"
        >
          Average
        </div>
        <div
          data-testid="line-Minimum"
        >
          Minimum
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TemperatureChart matches snapshot with empty data 1`] = `
<div
  class=""
>
  <div
    class="flex items-center mb-3 sm:mb-4"
  >
    <div
      class="flex items-center"
    >
      <svg
        class="w-4 h-4 sm:w-5 sm:h-5 text-gray-600 dark:text-gray-400 mr-2"
        fill="currentColor"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 2C10.34 2 9 3.34 9 5v6.5c-1.21.91-2 2.37-2 3.5a5 5 0 0 0 10 0c0-1.13-.79-2.59-2-3.5V5c0-1.66-1.34-3-3-3zm-1 3h2v2h-2V5zm0 3h2v2h-2V8zm0 3h2v2h-2v-2z"
        />
      </svg>
      <h3
        class="text-base sm:text-lg font-medium text-gray-900 dark:text-orange-200 bg-gradient-to-r from-orange-500 to-red-500 bg-clip-text text-transparent"
      >
        Temperature
      </h3>
    </div>
  </div>
  <div
    style="width: 100%; height: 250px; min-height: 200px;"
  >
    <div
      data-testid="responsive-container"
    >
      <div
        data-testid="line-chart"
      >
        <div
          data-testid="x-axis"
        >
          XAxis
        </div>
        <div
          data-testid="y-axis"
        >
          YAxis
        </div>
        <div
          data-testid="tooltip"
        >
          Tooltip
        </div>
        <div
          data-testid="legend"
        >
          Legend
        </div>
        <div
          data-testid="line-Maximum"
        >
          Maximum
        </div>
        <div
          data-testid="line-Average"
        >
          Average
        </div>
        <div
          data-testid="line-Minimum"
        >
          Minimum
        </div>
      </div>
    </div>
  </div>
</div>
`;
